test/e2e/cloud-provider-azure.go

```diff
@@ -0,0 +1,149 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	helmAction "helm.sh/helm/v3/pkg/action"
+	helmLoader "helm.sh/helm/v3/pkg/chart/loader"
+	helmCli "helm.sh/helm/v3/pkg/cli"
+	helmVals "helm.sh/helm/v3/pkg/cli/values"
+	helmGetter "helm.sh/helm/v3/pkg/getter"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// InstallCloudProviderAzureHelmChart installs the official cloud-provider-azure helm chart
+func InstallCloudProviderAzureHelmChart(ctx context.Context, input clusterctl.ApplyClusterTemplateAndWaitInput, result *clusterctl.ApplyClusterTemplateAndWaitResult) {
+	By("Waiting for workload cluster kubeconfig secret")
+	Eventually(func() error {
+		client := input.ClusterProxy.GetClient()
+		secret := &corev1.Secret{}
+		key := crclient.ObjectKey{
+			Name:      fmt.Sprintf("%s-kubeconfig", input.ConfigCluster.ClusterName),
+			Namespace: input.ConfigCluster.Namespace,
+		}
+		return client.Get(ctx, key, secret)
+	}, 20*time.Minute, 5*time.Second).Should(Succeed())
+	clusterProxy := input.ClusterProxy.GetWorkloadCluster(ctx, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
+	By("Waiting for nodes to come online indicating that the cluster is ready to accept work")
+	Eventually(func() error {
+		clientSet := clusterProxy.GetClientSet()
+		var runningNodes int
+		list, err := clientSet.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		for _, n := range list.Items {
+			if n.Status.Phase == corev1.NodeRunning {
+				runningNodes++
+			}
+		}
+		if runningNodes > 0 {
+			return nil
+		}
+		return err
+	}, 20*time.Minute, 5*time.Second).Should(Succeed())
+	kubeConfigPath := clusterProxy.GetKubeconfigPath()
+	os.Setenv("KUBECONFIG", kubeConfigPath)
+	By("Installing the correct version of cloud-provider-azure components via helm")
+	Eventually(func() error {
+		settings := helmCli.New()
+		actionConfig := new(helmAction.Configuration)
+		if err := actionConfig.Init(settings.RESTClientGetter(), "default", os.Getenv("HELM_DRIVER"), Logf); err != nil {
+			return err
+		}
+		i := helmAction.NewInstall(actionConfig)
+		i.RepoURL = "https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo"
+		i.ReleaseName = "cloud-provider-azure-oot"
```

Comment:
The release here could be a different patch version each time? It may be hard to determine if a change in the cloud-provider broke or something in capz.  Not sure there is a good option here as we want to be able to have the helm chart pick the right version based on k8s verion just calling it out.

Reply:
The name of the release is just a friendly, unique identifier for each "helm install" (and/or subsequent "helm upgrade"). Because the E2E flow is never helm install'ing more than once, I'm just choosing a static string (any string will do).